File: rubygems/pray-cli/lib/pray/cli/commands/init_prayer.rb
# frozen_string_literal: true

require "fileutils"

module Pray
  module CLI
    def prayer_init_command(name = nil, directory = nil)
      if File.exist?(manifest_path)
        scaffold_local_prayer(name, directory)
      else
        scaffold_standalone_package(name)
      end
    end

    def scaffold_local_prayer(name, directory)
      requested = name.to_s.strip
      package_name = LocalPrayer.validate_name!(requested.empty? ? DEFAULT_LOCAL_PRAYER_NAME : requested)
      source_name, source_directory = resolve_local_prayer_directory(directory)
      root = File.join(project_root, source_directory, package_name)
      prayspec_path = File.join(root, "#{package_name}.prayspec")
      raise Error.manifest("package spec already exists: #{prayspec_path}") if File.exist?(prayspec_path)
      raise Error.manifest("prayer directory already exists: #{root}") if File.exist?(root)

      FileUtils.mkdir_p(File.join(root, "exports"))
      File.write(prayspec_path, local_prayspec(source_name, package_name))
      File.write(File.join(root, "README.md"), "# #{package_name}\n")
      File.write(File.join(root, "exports", "#{package_name}.md"), "# #{package_name}\n")
      declare_local_prayer("#{source_name}/#{package_name}")
    end

    def scaffold_standalone_package(name)
      root = Dir.pwd
      package_name = if name.nil? || name.strip.empty?
        basename = File.basename(root)
        basename.strip.empty? ? "prayer-package" : basename
      else
        LocalPrayer.validate_name!(name)
      end
      prayspec_path = File.join(root, "#{package_name}.prayspec")
      raise Error.manifest("package spec already exists: #{prayspec_path}") if File.exist?(prayspec_path)

      File.write(prayspec_path, standalone_prayspec(package_name))
      readme = File.join(root, "README.md")
      File.write(readme, "# #{package_name}\n") unless File.exist?(readme)
      FileUtils.mkdir_p(File.join(root, "exports"))
    end

    def local_prayspec(source_name, name)
      <<~SPEC
        Package::Specification.new do |spec|
          spec.name = "#{source_name}/#{name}"
          spec.summary = "Describe this package"
          spec.files = ["README.md", "exports/#{name}.md"]
          spec.exports = {
            "#{name}" => {
              type: "fragment",
              path: "exports/#{name}.md"
            }
          }
        end
      SPEC
    end

    def standalone_prayspec(name)
      <<~SPEC
        Package::Specification.new do |spec|
          spec.name = "#{name}"
          spec.version = "0.1.0"
          spec.summary = "Describe this package"
          spec.files = ["README.md"]
          spec.exports = {}
        end
      SPEC
    end
  end
end

File: npmjs/pray-cli/src/cli/commands/init-prayer.ts
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { basename, join } from "node:path";
import { PrayError } from "../../errors.js";
import {
  DEFAULT_LOCAL_PRAYER_NAME,
  validateLocalPrayerName,
} from "../../local-prayer.js";
import { defaultManifestPath } from "../../lockfile/paths.js";
import { projectRoot } from "../invocation.js";
import {
  declareLocalPrayer,
  parsePrayerInitArguments,
  resolveLocalPrayerDirectory,
} from "./init-prayer-manifest.js";

export function runPrayerInit(argumentsList: string[] = []): void {
  const { name, directory } = parsePrayerInitArguments(argumentsList);
  if (existsSync(defaultManifestPath())) {
    scaffoldLocalPrayer(name, directory);
  } else {
    scaffoldStandalonePackage(name);
  }
}

function scaffoldLocalPrayer(name?: string, directory?: string): void {
  const packageName = validateLocalPrayerName(
    name && name.trim().length > 0 ? name : DEFAULT_LOCAL_PRAYER_NAME,
  );
  const source = resolveLocalPrayerDirectory(directory);
  const root = join(projectRoot(), source.directory, packageName);
  const prayspecPath = join(root, `${packageName}.prayspec`);
  if (existsSync(prayspecPath)) {
    throw PrayError.manifest(`package spec already exists: ${prayspecPath}`);
  }
  if (existsSync(root)) {
    throw PrayError.manifest(`prayer directory already exists: ${root}`);
  }
  mkdirSync(join(root, "exports"), { recursive: true });
  writeFileSync(prayspecPath, localPrayspec(source.name, packageName), "utf8");
  writeFileSync(join(root, "README.md"), `# ${packageName}\n`, "utf8");
  writeFileSync(
    join(root, "exports", `${packageName}.md`),
    `# ${packageName}\n`,
    "utf8",
  );
  declareLocalPrayer(`${source.name}/${packageName}`);
}

function scaffoldStandalonePackage(name?: string): void {
  const root = process.cwd();
  const packageName =
    name && name.trim().length > 0
      ? validateLocalPrayerName(name)
      : basename(root) || "prayer-package";
  const prayspecPath = join(root, `${packageName}.prayspec`);
  if (existsSync(prayspecPath)) {
    throw PrayError.manifest(`package spec already exists: ${prayspecPath}`);
  }
  writeFileSync(prayspecPath, standalonePrayspec(packageName), "utf8");
  const readmePath = join(root, "README.md");
  if (!existsSync(readmePath)) {
    writeFileSync(readmePath, `# ${packageName}\n`, "utf8");
  }
  mkdirSync(join(root, "exports"), { recursive: true });
}

function localPrayspec(sourceName: string, name: string): string {
  return `Package::Specification.new do |spec|
  spec.name = "${sourceName}/${name}"
  spec.summary = "Describe this package"
  spec.files = ["README.md", "exports/${name}.md"]
  spec.exports = {
    "${name}" => {
      type: "fragment",
      path: "exports/${name}.md"
    }
  }
end
`;
}

function standalonePrayspec(name: string): string {
  return `Package::Specification.new do |spec|
  spec.name = "${name}"
  spec.version = "0.1.0"
  spec.summary = "Describe this package"
  spec.files = ["README.md"]
  spec.exports = {}
end
`;
}
